data-products/src/main/scala/org/sunbird/analytics/sourcing/SourcingMetrics.scala

```diff
@@ -0,0 +1,173 @@
+      .set("spark.cassandra.input.consistency.level", readConsistencyLevel)
+    val spark = SparkSession.builder.config(sparkConf).getOrCreate()
+
+    generateSourcingMetrics(spark, config)
+  }
+
+  // $COVERAGE-ON$ Enabling scoverage for all other functions
+  def generateSourcingMetrics(spark: SparkSession, config: String)(implicit sc: SparkContext,fc: FrameworkContext): Unit = {
+    val conf = JSONUtils.deserialize[Map[String,AnyRef]](config)
+    val textbooks = TextBookUtils.getTextBooks(conf, RestUtil)
+    JobLogger.log(s"Fetched textbooks from druid ${textbooks.length}",None, Level.INFO)
+    import spark.implicits._
+
+    val textbookReportData = getTextbookInfo(textbooks, spark, config)
+    val textbookReports = textbookReportData._1.toDF()
+    val tenantInfo = getTenantInfo(RestUtil).toDF()
+    val report = textbookReports.join(tenantInfo,
+      textbookReports.col("channel") === tenantInfo.col("id"),"left")
+    generateSourcingReport(spark, textbookReportData._2.toDF(), report, config)
+  }
+
+  def generateSourcingReport(spark: SparkSession, contentdf: DataFrame, report: DataFrame, config: String): Unit = {
+    import spark.implicits._
+    val contentChapter = contentdf.groupBy("identifier","l1identifier")
+      .pivot(concat(lit("Number of "), col("contentType"))).agg(count("l1identifier"))
+    val contentTb = contentdf.groupBy("identifier")
+      .pivot(concat(lit("Number of "), col("contentType"))).agg(count("identifier"))
+    val configMap = JSONUtils.deserialize[Map[String,AnyRef]](config)
+    val reportConfig = configMap("modelParams").asInstanceOf[Map[String, AnyRef]]("reportConfig").asInstanceOf[Map[String, AnyRef]]
+    val reportPath = reportConfig.getOrElse("reportPath","sourcing").asInstanceOf[String]
+    val reportKey = reportConfig.getOrElse("reportKey","reports").asInstanceOf[String]
+    val storageConfig = getStorageConfig(reportKey, "")
+
+    val textbookReport = report.join(contentTb, Seq("identifier"),"left")
+      .drop("identifier","channel","id","chapters","l1identifier")
+      .distinct()
+      .orderBy('medium,split(split('grade,",")(0)," ")(1).cast("int"),'subject,'name)
+    saveReportToBlob(textbookReport, JSONUtils.serialize(reportConfig), storageConfig, "TextbookLevel", reportPath)
+
+    val chapterReport = report.join(contentChapter, Seq("identifier","l1identifier"),"left")
+      .drop("identifier","l1identifier","channel","id","totalChapters")
+      .orderBy('medium,split(split('grade,",")(0)," ")(1).cast("int"),'subject,'name,'chapters)
+    JobLogger.log(s"$jobName: extracted chapter and textbook reports", None, INFO)
+    saveReportToBlob(chapterReport, JSONUtils.serialize(reportConfig), storageConfig, "ChapterLevel", reportPath)
+  }
+
+  def getTextbookInfo(textbooks: List[TextbookData], spark: SparkSession, config: String)(implicit sc: SparkContext,fc: FrameworkContext): (List[TextbookReportResult],List[ContentReportResult]) = {
```

Comment:
Same as above. Pass in SparkSession, FrameworkContext and Config as implicit parameters